data/tutorials/lg_01_functional_programming.md

```diff
@@ -10,50 +10,327 @@ date: 2021-08-06T17:16:00-00:00
-types, pattern matching, type inference, nested functions) could exist in 
-a "super C" language. These are cool
-features that make your code concise, easy to read, and have
-fewer bugs; however, they actually have very little to do with functional
-programming. In fact, the reason 
-functional languages are so great is *not* because of functional
-programming but because we've been stuck with C-like languages for
-years, and in the meantime, cutting-edge programming has progressed
-considerably. So while we were writing
-`struct { int type; union { ... } }` for the umpteenth time, ML and
-Haskell programmers had safe variants and pattern matching on datatypes.
-While we were being careful to `free` all our `malloc`s, there have been
-languages with garbage collectors able to outperform hand coding since
-the 80s.
-
-Let's explore functional programming.
+
+At its core, and from its debut around 1985, OCaml is a [functional
+programming](https://en.wikipedia.org/wiki/Functional_programming) language. For
+instance, here is what was written about Caml, the ancestor of OCaml:
+
+<!--
+Les fonctions forment les constituants élémentaires des programmes en Caml. Un
+programme n’est rien d’autre qu’une collection de définitions de fonctions, suivie d’un
+appel à la fonction qui déclenche le calcul voulu.
+-->
+
+> Functions constitutes the basic components of Caml programs. A program is
+> nothing but a collection of function definitions, followed by a call to the
+> function which triggers the expected computation.
+
+Pierre Weis and Xavier Leroy, _Le langage Caml_, 1993
+
+OCaml has many other constituents and facets, such as:
+- Rich data-types
+- Pattern-matching
+- Type inference
+- Parametric polymorphism
+- Nested definitions
+- Exceptions
+- Garbage collected memory management
+- Modules and functors
+- Call-by-value evaluation
+- Native support for multiple programming paradigms
+  - Imperative
+  - Object-oriented
+- Domains-based support for multi-core hardware architectures
+- Distributed computing using effect-handlers
+
+Don't worry if those terms aren't making sense, yet. They are detailed in others
+documents. This is just a list of keywords.
+
+It is important to remember that all the features of OCaml are either built on
+top of functional programming or meant to support it. Most mainstream languages
+derive from the imperative programming paradigm, to which either structured or
+object-oriented programming where added. The feature set of OCaml, as sketched
+earlier, was pretty unique for many years. Nowadays, several languages have many
+of the features found in OCaml, but either their conception or their evolution
+goes in the opposite direction. OCaml started from functional programming while
+many others are going towards functional programming. This is not the place to
+detail the differences these paths entail. Let's just say OCaml developers and
+practitioners believe the outstanding level of reliability, performance and
```

Comment:
Could we just use "users"? "practitioners" has a medicine tone (maybe that's on purpose)